### DISCUSSION TITLE
Displaying a 2 x 3 image array with radio buttons

### DISCUSSION THREAD

**Discussion by Geoff-McK** [jspsych/jsPsych#1045] [Q&A] [answered]
Hi, I am trying to present six images on the screen, with 3 on a top row, and 3 on a bottom row, with radio buttons below each image. I can get the images where they need to be. However,  using the multi-choice plugin, I could not figure how to manipulate the position of the radio buttons. I managed to spread the radio buttons out, but that was about it. I also will need to place a separate radio button (that is not associated with any of the images). I figured that I could simply create an empty prompt, but I don't know how I would manipulate its position. 

Any feedback would be greatly appreciated. 

Thanks.

**Comment by becky-gilbert** [comment]
Hi @Geoff-McK, were you able to get any further with this? If not could you share your code?

As you may have already figured out, you can add CSS to manipulate the formatting/positioning of all of your radio buttons using this CSS selector:
```css
input[type="radio"]
```
And you can adjust each radio button individually using its ID, e.g.
```css
input[type="radio"]#jspsych-survey-multi-choice-response-0-0
```

**Comment by Geoff-McK** [comment]
Hi Becky,

Thank you (again) for your help. Because I am so new to online coding (e.g., HTML, Javascript, etc.), I don't think I have enough top-down knowledge to interpret your advice. 

I was able to use your advice to accomplish my goal. However, I don't know if I did what you were suggesting. I was hoping that I could get some more specific advice, if possible. Whatever advice you could give would be greatly appreciated.

My code from the jspsych-survey-multi-choice plugin is below (I renamed it to "jspsych-lineup-multi-choice"). 

In the for loop for creating radio buttons, I removed the bit of code that adds a radio button container. Instead, I moved this bit of code outside of the for loop, copy and pasted it as many times as I needed it, and added html code that specified the position and label of each option, which was specific to each radio button. This is likely a very inefficient way to do this, and I did not use the "#" sign, so I'm thinking that I probably didn't do what you were suggesting. 

// add multiple-choice questions
      for (var i = 0; i < trial.questions.length; i++) {
        
        // get question based on question_order
        var question = trial.questions[question_order[i]];
        var question_id = question_order[i];
        
        // create question container
        var question_classes = ['jspsych-lineup-multi-choice-question'];
        if (question.horizontal) {
          question_classes.push('jspsych-lineup-multi-choice-horizontal');
        }
  
        html += '<div id="jspsych-lineup-multi-choice-'+question_id+'" class="'+question_classes.join(' ')+'"  data-name="'+question.name+'">';
  
        // add question text
        html += '<p class="jspsych-lineup-multi-choice-text lineup-multi-choice">' + question.prompt 
        if(question.required){
          html += "<span class='required'></span>";
        }
        html += '</p>';
  
        // create option radio buttons
        for (var j = 0; j < question.options.length; j++) {
          // add label and question text
          var option_id_name = "jspsych-lineup-multi-choice-option-"+question_id+"-"+j;
          var input_name = 'jspsych-lineup-multi-choice-response-'+question_id;
          var input_id = 'jspsych-lineup-multi-choice-response-'+question_id+'-'+j;


          var required_attr = question.required ? 'required' : '';
          console.log("question.required: " + question.required);
          
        }


        
        // add radio button container

        html += '<div style="position: fixed; top: 660px; left: 175px;" id="jspsych-lineup-multi-choice-option-0-6" class="jspsych-lineup-multi-choice-option">';
        html += '<label class="jspsych-lineup-multi-choice-text" for="jspsych-lineup-multi-choice-response-0-6">ABSENT</label>';
        html += '<input type="radio" name="jspsych-lineup-multi-choice-response-0" id="jspsych-lineup-multi-choice-response-0-6" value="ABSENT"></input>';
        html += '</div>';
        html += '</div>';

        html += '<div style="position: fixed; top: 250px; left: 100px;" id="jspsych-lineup-multi-choice-option-0-0" class="jspsych-lineup-multi-choice-option">';
        html += '<label class="jspsych-lineup-multi-choice-text" for="jspsych-lineup-multi-choice-response-0-0">1</label>';
        html += '<input type="radio" name="jspsych-lineup-multi-choice-response-0" id="jspsych-lineup-multi-choice-response-0-0" value="1" ></input>';
        html += '</div>';

       html += '<div style="position: fixed; top: 250px; left: 300px;" id="jspsych-lineup-multi-choice-option-0-1" class="jspsych-lineup-multi-choice-option">';
        html += '<label class="jspsych-lineup-multi-choice-text" for="jspsych-lineup-multi-choice-response-0-1">2</label>';
        html += '<input type="radio" name="jspsych-lineup-multi-choice-response-0" id="jspsych-lineup-multi-choice-response-0-1" value="2" ></input>';
        html += '</div>';

        html += '<div style="position: fixed; top: 250px; left: 500px;" id="jspsych-lineup-multi-choice-option-0-2" class="jspsych-lineup-multi-choice-option">';
        html += '<label class="jspsych-lineup-multi-choice-text" for="jspsych-lineup-multi-choice-response-0-2">3</label>';
        html += '<input type="radio" name="jspsych-lineup-multi-choice-response-0" id="jspsych-lineup-multi-choice-response-0-2" value="3" ></input>';
        html += '</div>';

        html += '<div style="position: fixed; top: 600px; left: 100px;" id="jspsych-lineup-multi-choice-option-0-3" class="jspsych-lineup-multi-choice-option">';
        html += '<label class="jspsych-lineup-multi-choice-text" for="jspsych-lineup-multi-choice-response-0-3">4</label>';
        html += '<input type="radio" name="jspsych-lineup-multi-choice-response-0" id="jspsych-lineup-multi-choice-response-0-3" value="4" ></input>';
        html += '</div>';

        html += '<div style="position: fixed; top: 600px; left: 300px;" id="jspsych-lineup-multi-choice-option-0-4" class="jspsych-lineup-multi-choice-option">';
        html += '<label class="jspsych-lineup-multi-choice-text" for="jspsych-lineup-multi-choice-response-0-4">5</label>';
        html += '<input type="radio" name="jspsych-lineup-multi-choice-response-0" id="jspsych-lineup-multi-choice-response-0-5" value="5" ></input>';
        html += '</div>';

        html += '<div style="position: fixed; top: 600px; left: 500px;" id="jspsych-lineup-multi-choice-option-0-5" class="jspsych-lineup-multi-choice-option">';
        html += '<label class="jspsych-lineup-multi-choice-text" for="jspsych-lineup-multi-choice-response-0-5">6</label>';
        html += '<input type="radio" name="jspsych-lineup-multi-choice-response-0" id="jspsych-lineup-multi-choice-response-0-5" value="6" '+required_attr+'></input>';
        html += '</div>';
        


      }

**Reply by becky-gilbert** [reply]
Ah ok, I didn't realize that you were creating a modified plugin. In that case you should ignore my earlier advice, which was about how to reposition elements on the page using the standard 'survey-multi-choice' plugin and custom CSS. 

I think it makes sense to create a custom modified plugin, as you have done here. One suggestion I have is to add the radio button HTML to the larger HTML string _inside_ the for loop that loops through each radio button option, i.e. the one starting with:
```javascript
// create option radio buttons
for (var j = 0; j < question.options.length; j++) {
  ...
```
You can use the `j` value to define the things that change across the different radio button options, like the `id` and `top`/`left` CSS values. You can also add the first 'absent' option by putting in an if statement that adds the absent option HTML to the larger HTML string when the value of `j` is 0. 

I'll see if I can put something together for you and post it here soon.

**Reply by becky-gilbert** [reply]
Could you show me a simplified example of how you're setting up the lineup trials? I'm wondering how you're passing the images into the plugin. Are you using an array of image file paths for the `options` parameter, like this?
```javascript
var image_options = [
    "jspsych-6.1.0/examples/img/happy_face_1.jpg", 
    "jspsych-6.1.0/examples/img/happy_face_2.jpg", 
    "jspsych-6.1.0/examples/img/happy_face_3.jpg", 
    "jspsych-6.1.0/examples/img/happy_face_4.jpg", 
    "jspsych-6.1.0/examples/img/sad_face_1.jpg",
    "jspsych-6.1.0/examples/img/sad_face_2.jpg"
];

var lineup_trial = {
    type: 'lineup-multi-choice',
    questions: [
        {prompt: "", name: 'lineup', options: image_options, required: true}
    ]
};
```

**Comment by Geoff-McK** [comment]
Hi Becky, 

Thank you again for all of your help. Below is a simplified version of my code. Hopefully, it is not too burdensome. The code looked weird when I tried to write it so I had to modify it a bit.

var survey_page2 = {
    type: 'survey-text',
    questions: [
        {prompt: "What is your subject number?", columns: 4, required: true, name: 'subject'}
    ],
    data: {task: 'data-input'},
       on_finish: function(data){
# use the user input to extract the variables "target" and "video_order"
        if(video_order==0){
          if(target==0){
            lineup = ['CarTheftFiller1.jpg','CarTheftFiller2.jpg','CarTheftFiller3.jpg','CarTheftFiller4.jpg','CarTheftFiller5.jpg','CarTheftFiller6.jpg'];
        } else {
            lineup = ['CarTheftFiller1.jpg','CarTheftFiller2.jpg','CarTheftFiller3.jpg','CarTheftFiller4.jpg','CarTheftFiller5.jpg','CarTheftPerp.jpg'];
        }

        } else if (video_order==1){
          if(target===0){
            lineup = ['AutopartCrimeFiller3.jpg','AutopartCrimeFiller10.jpg','AutopartCrimeFiller11.jpg','AutopartCrimeFiller12.jpg','AutopartCrimeFiller14.jpg','AutopartCrimeFiller5.jpg'];
        } else {
            lineup = ['AutopartCrimeFiller3.jpg','AutopartCrimeFiller10.jpg','AutopartCrimeFiller11.jpg','AutopartCrimeFiller12.jpg','AutopartCrimeFiller14.jpg','AutopartCrimePerp.jpg'];
        }
        }

lineup = jsPsych.randomization.repeat(lineup, 1);
        
      }    
}; 

var page_1_options = ["ABSENT","1","2","3","4","5","6",];

var multi_choice_block_horizontal = {
    type: 'lineup-multi-choice',
    questions: [        
      {prompt: '<div <p </p>', options: page_1_options, required: true, horizontal: true},
],
    button_label: 'SUBMIT',
    preamble: function() {
    var preamble_html = 
    "<div style='position: fixed; top: 0px; left: 20px;'><p>If you see the person who committed the crime in the previous</p>" +
      "<div style='position: fixed; top: 20px; left: 20px;'><p>video, type in the number that corresponds to their photo.</p>" +
      "<div style='position: fixed; top: 60px; left: 20px;'><p>If the person who committed the crime is <strong>NOT </strong> one of the</p>" +
      "<div style='position: fixed; top: 80px; left: 20px;'><p>faces that you see, type ABSENT</p>"

// + div p style = 'position: fixed; top: 150px; left: 0px; width: 200px;' image source attribute =" + the 0th index of lineup +" height=162.75 width=175> image p  +"  

// do this for all pictures

    return preamble_html;
    }
 };

**Reply by becky-gilbert** [reply]
Side note: to get around the problems with your code formatting on GitHub, at the start of your whole code block, you can add three backticks followed by the word javascript. And then to end your code block, add three more backticks. Like this:

\```javascript
var foo = bar;
function more_code() {
  return true;
}
\```

Then, if you click the 'Preview' tab at the top of your comment box, you should see that everything inside of the first and second set of backticks is properly formatted with syntax highlighting, like this:

```javascript
var foo = bar;
function more_code() {
  return true;
}
```

**Comment by Geoff-McK** [comment]
Thank you! Sorry for all of the annoying emails that you probably received while I was trying to edit my post. 

Here is the code that I had been trying to write:

```javascript 
var survey_page2 = {
    type: 'survey-text',
    questions: [
        {prompt: "What is your subject number?", columns: 4, required: true, name: 'subject'}
    ],
    data: {task: 'data-input'},
       on_finish: function(data){
       
      // use user input to extract variables "target" and "video_order"

        if(video_order==0){
          if(target==0){
            lineup = ['CarTheftFiller1.jpg','CarTheftFiller2.jpg','CarTheftFiller3.jpg','CarTheftFiller4.jpg','CarTheftFiller5.jpg','CarTheftFiller6.jpg'];
        } else {
            lineup = ['CarTheftFiller1.jpg','CarTheftFiller2.jpg','CarTheftFiller3.jpg','CarTheftFiller4.jpg','CarTheftFiller5.jpg','CarTheftPerp.jpg'];
        }

        } else if (video_order==1){
          if(target===0){
            lineup = ['AutopartCrimeFiller3.jpg','AutopartCrimeFiller10.jpg','AutopartCrimeFiller11.jpg','AutopartCrimeFiller12.jpg','AutopartCrimeFiller14.jpg','AutopartCrimeFiller5.jpg'];
        } else {
            lineup = ['AutopartCrimeFiller3.jpg','AutopartCrimeFiller10.jpg','AutopartCrimeFiller11.jpg','AutopartCrimeFiller12.jpg','AutopartCrimeFiller14.jpg','AutopartCrimePerp.jpg'];
        }
        }

lineup = jsPsych.randomization.repeat(lineup, 1);


      }    
}; 



var page_1_options = ["ABSENT","1","2","3","4","5","6",];


var multi_choice_block_horizontal = {
    type: 'lineup-multi-choice',
    questions: [        
      {prompt: '<div <p </p>', options: page_1_options, required: true, horizontal: true},
],
    button_label: 'SUBMIT',
    preamble: function() {
    var preamble_html = 
    "<div style='position: fixed; top: 0px; left: 20px;'><p>If you see the person who committed the crime in the previous</p>" +
      "<div style='position: fixed; top: 20px; left: 20px;'><p>video, type in the number that corresponds to their photo.</p>" +
      "<div style='position: fixed; top: 60px; left: 20px;'><p>If the person who committed the crime is <strong>NOT </strong> one of the</p>" +
      "<div style='position: fixed; top: 80px; left: 20px;'><p>faces that you see, type ABSENT</p>" +

  "<div <p style = 'position: fixed; top: 150px; left: 0px; width: 200px;'><img src=" + lineup[0] +" height=162.75 width=175></img></p>  " +  

  "<div <p style = 'position: fixed; top: 150px; left: 200px; width: 200px;'><img src=" + lineup[1] +" height=162.75 width=175></img></p>" + 

  "<div <p style = 'position: fixed; top: 150px; left: 400px; width: 200px;'><img src=" + lineup[2] +" height=162.75 width=175></img></p>  " +

  "<div <p style = 'position: fixed; top: 380px; left: 0px; width: 200px'><img src=" + lineup[3] +" height=162.75 width=175</img></p> " + 

  "<div <p style = 'position: fixed; top: 380px; left: 200px; width: 200px'><img src=" + lineup[4] +" height=162.75 width=175></img></p> " + 

  "<div <p style = 'position: fixed; top: 380px; left: 400px; width: 200px'><img src=" + lineup[5] +" height=162.75 width=175></img></p> "
    return preamble_html;
    }
 };

**Comment by becky-gilbert** [comment, marked as answer]
If you want the images to appear above each multiple choice radio button and label, then I would suggest passing the images into the plugin and adding them to the page along with the HTML for the radio buttons, because that will make the positioning a lot easier. So what I did was use your image array for the question's `options` parameter in your trial definition, so that inside the plugin, I could add each image in with the radio buttons.  Then for the number labels, you can just add those using the index value for the `options` image array (instead of passing those labels in as an array). I added 1 to the array index value so that your option numbering starts at 1 instead of 0.

I also "hard coded" added the ABSENT option into the plugin code so that you don't need to pass it into the plugin (this assumes that you always want this option to appear - not sure if that's the case). 

I also tweaked the positioning by changing some of your top/left positioning values, and added fixed positioning for the submit button so that it appears below the image options. You can change the positioning by playing around with changing the top/left values in the plugin.

Finally, the data wasn't saving because inside the form's submit event function, for each question, the code was still looking for the old question IDs (`#jspscyh-survey-multi-choice-0` etc.) but those don't exist anymore. So I changed that to match your new question IDs (`#jspsych-lineup-multi-choice-0`).

Here's how I set up the trial in the HTML file:
```javascript
    var lineup = [
        "jspsych-6.1.0/examples/img/happy_face_1.jpg", 
        "jspsych-6.1.0/examples/img/happy_face_2.jpg", 
        "jspsych-6.1.0/examples/img/happy_face_3.jpg", 
        "jspsych-6.1.0/examples/img/happy_face_4.jpg", 
        "jspsych-6.1.0/examples/img/sad_face_1.jpg",
        "jspsych-6.1.0/examples/img/sad_face_2.jpg"
    ];

    var multi_choice_block_horizontal = {
        type: 'lineup-multi-choice',
        questions: [        
            {prompt: '', options: lineup, required: true, horizontal: true},
        ],
        button_label: 'SUBMIT',
        preamble: "<div style='position: fixed; top: 0px; left: 100px;'><p>If you see the person who committed the crime in the previous</p>" +
            "<div style='position: fixed; top: 20px; left: 100px;'><p>video, type in the number that corresponds to their photo.</p>" +
            "<div style='position: fixed; top: 60px; left: 100px;'><p>If the person who committed the crime is <strong>NOT </strong> one of the</p>" +
            "<div style='position: fixed; top: 80px; left: 100px;'><p>faces that you see, select ABSENT</p>"
    };

    jsPsych.init({
        timeline: [multi_choice_block_horizontal],
        preload_images: lineup,
        on_finish: function() {
            jsPsych.data.displayData();
        }
    });
```

And here's the plugin code, starting from `plugin.trial`:
```javascript
plugin.trial = function(display_element, trial) {
    var plugin_id_name = "jspsych-survey-multi-choice";

    var html = "";

    // inject CSS for trial
    html += '<style id="jspsych-survey-multi-choice-css">';
    html += ".jspsych-survey-multi-choice-question { margin-top: 2em; margin-bottom: 2em; text-align: left; }"+
      ".jspsych-survey-multi-choice-text span.required {color: darkred;}"+
      ".jspsych-survey-multi-choice-horizontal .jspsych-survey-multi-choice-text {  text-align: center;}"+
      ".jspsych-survey-multi-choice-option { line-height: 2; }"+
      ".jspsych-survey-multi-choice-horizontal .jspsych-survey-multi-choice-option {  display: inline-block;  margin-left: 1em;  margin-right: 1em;  vertical-align: top;}"+
      "label.jspsych-survey-multi-choice-text input[type='radio'] {margin-right: 1em;}";
    html += '</style>';

    // show preamble text
    if(trial.preamble !== null){
      html += '<div id="jspsych-survey-multi-choice-preamble" class="jspsych-survey-multi-choice-preamble">'+trial.preamble+'</div>';
    }

    // form element
    html += '<form id="jspsych-survey-multi-choice-form">';
    
    // generate question order. this is randomized here as opposed to randomizing the order of trial.questions
    // so that the data are always associated with the same question regardless of order
    var question_order = [];
    for(var i=0; i<trial.questions.length; i++){
      question_order.push(i);
    }
    if(trial.randomize_question_order){
      question_order = jsPsych.randomization.shuffle(question_order);
    }

    // add multiple-choice questions
    for (var i = 0; i < trial.questions.length; i++) {

        // get question based on question_order
        var question = trial.questions[question_order[i]];
        var question_id = question_order[i];
        
        // create question container
        var question_classes = ['jspsych-lineup-multi-choice-question'];
        if (question.horizontal) {
            question_classes.push('jspsych-lineup-multi-choice-horizontal');
        }

        html += '<div id="jspsych-lineup-multi-choice-'+question_id+'" class="'+question_classes.join(' ')+'"  data-name="'+question.name+'">';

        // add question text
        html += '<p class="jspsych-lineup-multi-choice-text lineup-multi-choice">' + question.prompt 
        if (question.required) {
            html += "<span class='required'></span>";
        }
        html += '</p>';

        var required_attr = question.required ? 'required' : '';

        // create option radio buttons
        for (var j = 0; j < question.options.length; j++) {

            // for each response option, add radio button container, image and radio button/label
            // dynamically set values for label and input based on the option index j
            var option_id_name = "jspsych-lineup-multi-choice-option-"+question_id+"-"+j;
            var input_name = 'jspsych-lineup-multi-choice-response-'+question_id;
            var input_id = 'jspsych-lineup-multi-choice-response-'+question_id+'-'+j;

            var left = '100';
            if (j == 1 | j == 4) {
                left = '300';
            } else if (j == 2 | j == 5) {
                left = '500';
            }
            var top = '250';
            if (j < 3) {
                top = '600';
            } 
            
            // create the HTML for this response option
            html += '<div style="position: fixed; top: '+top+'px; left: '+left+'px;" id="'+option_id_name+'" class="jspsych-lineup-multi-choice-option">';
            html += '<img src="'+question.options[j]+'" height="162.75px"; width="175px";/>';
            html += '<p><label class="jspsych-lineup-multi-choice-text" for="'+input_id+'">'+(j+1).toString()+'</label>';
            html += '<input type="radio" name="'+input_name+'" id="'+input_id+'" value="'+(j+1).toString()+'" '+required_attr+' ></input></p>';
            html += '</div>';
        }
        // add the ABSENT option
        html += '<div style="position: fixed; top: 850px; left: 150px;" id="jspsych-lineup-multi-choice-option-0-6" class="jspsych-lineup-multi-choice-option">';
        html += '<label class="jspsych-lineup-multi-choice-text" for="jspsych-lineup-multi-choice-response-0-6">ABSENT</label>';
        html += '<input type="radio" name="jspsych-lineup-multi-choice-response-0" id="jspsych-lineup-multi-choice-response-0-6" value="ABSENT"></input>';
        html += '</div>';
        html += '</div>';
    }
    
    // add submit button
    html += '<div style="position: fixed; top: 850px; left: 350px; id="jspsych-lineup-multi-choice-submit">'
    html += '<input type="submit" id="'+plugin_id_name+'-next" class="'+plugin_id_name+' jspsych-btn"' + (trial.button_label ? ' value="'+trial.button_label + '"': '') + '></input>';
    html += '</div></form>';

    // render
    display_element.innerHTML = html;

    document.querySelector('form').addEventListener('submit', function(event) {
      event.preventDefault();
      // measure response time
      var endTime = performance.now();
      var response_time = endTime - startTime;

      // create object to hold responses
      var question_data = {};
      for(var i=0; i<trial.questions.length; i++){
        var match = display_element.querySelector('#jspsych-lineup-multi-choice-'+i);
        var id = "Q" + i;
        if(match.querySelector("input[type=radio]:checked") !== null){
          var val = match.querySelector("input[type=radio]:checked").value;
        } else {
          var val = "";
        }
        var obje = {};
        var name = id;
        if(match.attributes['data-name'].value !== ''){
          name = match.attributes['data-name'].value;
        }
        obje[name] = val;
        Object.assign(question_data, obje);
      }
      // save data
      var trial_data = {
        "rt": response_time,
        "responses": JSON.stringify(question_data),
        "question_order": JSON.stringify(question_order)
      };
      display_element.innerHTML = '';

      // next trial
      jsPsych.finishTrial(trial_data);
    });

    var startTime = performance.now();
  };
```
And here's what your trial looks like for me:

![lineup](https://user-images.githubusercontent.com/9041788/94608959-06873f00-0253-11eb-8257-b6ae00ca9c27.PNG)
